File: frontend/src/services/knowledge.ts
import { api } from './api';

export interface Paper {
  id: string;
  doi: string | null;
  pmid: string | null;
  arxiv_id: string | null;
  title: string;
  authors: string[];
  journal: string | null;
  publication_date: string | null;
  abstract: string | null;
  pdf_url: string | null;
  notes: string | null;
  read_status: 'unread' | 'reading' | 'read';
  rating: number | null;
  ai_summary: string | null;
  ai_key_findings: string[] | null;
  organization_id: string;
  added_by_id: string;
  created_at: string;
  updated_at: string;
}

export interface Collection {
  id: string;
  name: string;
  description: string | null;
  color: string | null;
  is_shared: boolean;
  paper_count: number;
  organization_id: string;
  created_by_id: string;
  created_at: string;
  updated_at: string;
}

export interface PaperHighlight {
  id: string;
  paper_id: string;
  text: string;
  color: string;
  note: string | null;
  page_number: number | null;
  position_data: Record<string, unknown> | null;
  created_by_id: string;
  created_at: string;
}

export interface PaperLink {
  id: string;
  paper_id: string;
  linked_entity_type: 'project' | 'task' | 'document';
  linked_entity_id: string;
  link_type: 'reference' | 'citation' | 'related';
  notes: string | null;
  created_by_id: string | null;
  created_at: string;
}

export interface CreatePaperRequest {
  doi?: string;
  pmid?: string;
  title: string;
  authors?: string[];
  journal?: string;
  publication_date?: string;
  abstract?: string;
  pdf_url?: string;
  organization_id: string;
}

export interface UpdatePaperRequest {
  title?: string;
  authors?: string[];
  journal?: string;
  publication_date?: string;
  abstract?: string;
  pdf_url?: string;
  notes?: string;
  read_status?: 'unread' | 'reading' | 'read';
  rating?: number;
}

export interface CreateCollectionRequest {
  name: string;
  description?: string;
  color?: string;
  organization_id: string;
}

export interface UpdateCollectionRequest {
  name?: string;
  description?: string;
  color?: string;
  is_shared?: boolean;
}

export interface CreateHighlightRequest {
  paper_id: string;
  text: string;
  color?: string;
  note?: string;
  page_number?: number;
  position_data?: Record<string, unknown>;
}

export interface CreatePaperLinkRequest {
  paper_id: string;
  linked_entity_type: 'project' | 'task' | 'document';
  linked_entity_id: string;
  link_type?: 'reference' | 'citation' | 'related';
  notes?: string;
}

// Papers
export async function createPaper(data: CreatePaperRequest): Promise<Paper> {
  const response = await api.post<Paper>('/knowledge/papers', data);
  return response.data;
}

export async function getPapers(
  organizationId: string,
  params?: {
    search?: string;
    read_status?: string;
    collection_id?: string;
    skip?: number;
    limit?: number;
  }
): Promise<Paper[]> {
  const response = await api.get<Paper[]>('/knowledge/papers', {
    params: { organization_id: organizationId, ...params },
  });
  return response.data || [];
}

export async function getPaper(paperId: string): Promise<Paper> {
  const response = await api.get<Paper>(`/knowledge/papers/${paperId}`);
  return response.data;
}

export async function updatePaper(paperId: string, data: UpdatePaperRequest): Promise<Paper> {
  const response = await api.patch<Paper>(`/knowledge/papers/${paperId}`, data);
  return response.data;
}

export async function deletePaper(paperId: string): Promise<void> {
  await api.delete(`/knowledge/papers/${paperId}`);
}

export async function importPaperByDOI(
  doi: string,
  organizationId: string
): Promise<Paper> {
  const response = await api.post<Paper>('/knowledge/papers/import/doi', {
    doi,
    organization_id: organizationId,
  });
  return response.data;
}

export async function importPaperByPMID(
  pmid: string,
  organizationId: string
): Promise<Paper> {
  const response = await api.post<Paper>('/knowledge/papers/import/pmid', {
    pmid,
    organization_id: organizationId,
  });
  return response.data;
}

// Collections
export async function createCollection(data: CreateCollectionRequest): Promise<Collection> {
  const response = await api.post<Collection>('/knowledge/collections', data);
  return response.data;
}

export async function getCollections(
  organizationId: string,
  params?: { skip?: number; limit?: number }
): Promise<Collection[]> {
  const response = await api.get<Collection[]>('/knowledge/collections', {
    params: { organization_id: organizationId, ...params },
  });
  return response.data || [];
}

export async function getCollection(collectionId: string): Promise<Collection> {
  const response = await api.get<Collection>(`/knowledge/collections/${collectionId}`);
  return response.data;
}

export async function updateCollection(
  collectionId: string,
  data: UpdateCollectionRequest
): Promise<Collection> {
  const response = await api.patch<Collection>(
    `/knowledge/collections/${collectionId}`,
    data
  );
  return response.data;
}

export async function deleteCollection(collectionId: string): Promise<void> {
  await api.delete(`/knowledge/collections/${collectionId}`);
}

export async function addPaperToCollection(
  collectionId: string,
  paperId: string
): Promise<void> {
  await api.post(`/knowledge/collections/${collectionId}/papers/${paperId}`);
}

export async function removePaperFromCollection(
  collectionId: string,
  paperId: string
): Promise<void> {
  await api.delete(`/knowledge/collections/${collectionId}/papers/${paperId}`);
}

export async function getCollectionPapers(
  collectionId: string,
  params?: { skip?: number; limit?: number }
): Promise<Paper[]> {
  const response = await api.get<Paper[]>(
    `/knowledge/collections/${collectionId}/papers`,
    { params }
  );
  return response.data || [];
}

// Highlights
export async function createHighlight(data: CreateHighlightRequest): Promise<PaperHighlight> {
  const response = await api.post<PaperHighlight>(
    `/knowledge/papers/${data.paper_id}/highlights`,
    data
  );
  return response.data;
}

export async function getPaperHighlights(paperId: string): Promise<PaperHighlight[]> {
  const response = await api.get<PaperHighlight[]>(
    `/knowledge/papers/${paperId}/highlights`
  );
  return response.data || [];
}

export async function deleteHighlight(highlightId: string): Promise<void> {
  await api.delete(`/knowledge/highlights/${highlightId}`);
}

// Paper Links
export async function createPaperLink(data: CreatePaperLinkRequest): Promise<PaperLink> {
  const response = await api.post<PaperLink>('/knowledge/paper-links', data);
  return response.data;
}

export async function getPaperLinks(paperId: string): Promise<PaperLink[]> {
  const response = await api.get<PaperLink[]>(`/knowledge/papers/${paperId}/links`);
  return response.data || [];
}

export async function deletePaperLink(linkId: string): Promise<void> {
  await api.delete(`/knowledge/paper-links/${linkId}`);
}

// Get papers linked to a project
export async function getProjectPapers(projectId: string): Promise<Paper[]> {
  const response = await api.get<Paper[]>(`/knowledge/projects/${projectId}/papers`);
  return response.data || [];
}

// Link a paper to a project
export async function linkPaperToProject(paperId: string, projectId: string, notes?: string): Promise<PaperLink> {
  return createPaperLink({
    paper_id: paperId,
    linked_entity_type: 'project',
    linked_entity_id: projectId,
    link_type: 'reference',
    notes,
  });
}

File: backend/researchhub/models/knowledge.py
"""Knowledge management models - papers, collections, and references."""

from datetime import datetime, date
from typing import TYPE_CHECKING
from uuid import UUID

from sqlalchemy import Boolean, Date, DateTime, ForeignKey, Integer, String, Text, UniqueConstraint
from sqlalchemy.dialects.postgresql import ARRAY, JSONB, UUID as PGUUID, TSVECTOR
from sqlalchemy.orm import Mapped, mapped_column, relationship

from researchhub.db.base import BaseModel

if TYPE_CHECKING:
    from researchhub.models.organization import Organization
    from researchhub.models.user import User


class Paper(BaseModel):
    """Research paper in the knowledge library."""

    __tablename__ = "papers"

    # External identifiers
    doi: Mapped[str | None] = mapped_column(String(255), nullable=True, index=True)
    pmid: Mapped[str | None] = mapped_column(String(50), nullable=True, index=True)
    arxiv_id: Mapped[str | None] = mapped_column(String(50), nullable=True, index=True)

    # Basic metadata
    title: Mapped[str] = mapped_column(String(1000), nullable=False)
    authors: Mapped[list[str]] = mapped_column(
        ARRAY(String), nullable=False, server_default="{}"
    )
    journal: Mapped[str | None] = mapped_column(String(500), nullable=True)
    publication_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    publication_year: Mapped[int | None] = mapped_column(Integer, nullable=True)

    # Content
    abstract: Mapped[str | None] = mapped_column(Text, nullable=True)
    keywords: Mapped[list[str]] = mapped_column(
        ARRAY(String), nullable=False, server_default="{}"
    )

    # Files
    pdf_url: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    pdf_file_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True), nullable=True
    )  # Reference to attachment

    # AI-generated content
    ai_summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    ai_key_findings: Mapped[list[str]] = mapped_column(
        ARRAY(String), nullable=False, server_default="{}"
    )
    ai_methodology: Mapped[str | None] = mapped_column(Text, nullable=True)
    ai_processed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    # User notes
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Organization and ownership
    organization_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("organizations.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    added_by_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )

    # Reading status
    read_status: Mapped[str] = mapped_column(
        String(50), nullable=False, default="unread"
    )  # unread, reading, read
    read_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    # User rating (1-5 stars)
    rating: Mapped[int | None] = mapped_column(Integer, nullable=True)

    # Tags for organization
    tags: Mapped[list[str]] = mapped_column(
        ARRAY(String), nullable=False, server_default="{}"
    )

    # Citation information
    citation_count: Mapped[int | None] = mapped_column(Integer, nullable=True)
    bibtex: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Full-text search vector
    search_vector: Mapped[str | None] = mapped_column(TSVECTOR, nullable=True)

    # Metadata from external sources
    external_metadata: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)

    # Relationships
    organization: Mapped["Organization"] = relationship("Organization")
    added_by: Mapped["User | None"] = relationship("User")
    collection_memberships: Mapped[list["CollectionPaper"]] = relationship(
        "CollectionPaper", back_populates="paper", lazy="selectin"
    )

    def __repr__(self) -> str:
        try:
            return f"<Paper {self.title[:50]}>"
        except Exception:
            return f"<Paper id={self.id}>"


class Collection(BaseModel):
    """Collection of papers for organization."""

    __tablename__ = "collections"

    name: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Collection type
    is_smart: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)

    # For smart collections - filter criteria
    filter_criteria: Mapped[dict | None] = mapped_column(JSONB, nullable=True)

    # Visual customization
    color: Mapped[str | None] = mapped_column(String(7), nullable=True)
    icon: Mapped[str | None] = mapped_column(String(50), nullable=True)

    # Ownership
    organization_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("organizations.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    created_by_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )

    # Visibility
    visibility: Mapped[str] = mapped_column(
        String(50), nullable=False, default="private"
    )  # private, team, organization

    # Relationships
    organization: Mapped["Organization"] = relationship("Organization")
    created_by: Mapped["User | None"] = relationship("User")
    paper_memberships: Mapped[list["CollectionPaper"]] = relationship(
        "CollectionPaper", back_populates="collection", lazy="selectin"
    )

    def __repr__(self) -> str:
        try:
            return f"<Collection {self.name}>"
        except Exception:
            return f"<Collection id={self.id}>"


class CollectionPaper(BaseModel):
    """Many-to-many relationship between collections and papers."""

    __tablename__ = "collection_papers"
    __table_args__ = (
        UniqueConstraint("collection_id", "paper_id", name="uq_collection_paper"),
    )

    collection_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("collections.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    paper_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("papers.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # Position in collection (for manual ordering)
    position: Mapped[int] = mapped_column(Integer, nullable=False, default=0)

    # Who added it
    added_by_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )

    # Relationships
    collection: Mapped["Collection"] = relationship(
        "Collection", back_populates="paper_memberships"
    )
    paper: Mapped["Paper"] = relationship("Paper", back_populates="collection_memberships")
    added_by: Mapped["User | None"] = relationship("User")

    def __repr__(self) -> str:
        return f"<CollectionPaper coll={self.collection_id} paper={self.paper_id}>"


class PaperHighlight(BaseModel):
    """Highlight or annotation on a paper."""

    __tablename__ = "paper_highlights"

    paper_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("papers.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    user_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )

    # Highlight content
    highlighted_text: Mapped[str] = mapped_column(Text, nullable=False)
    note: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Position in PDF (page, coordinates)
    position_data: Mapped[dict] = mapped_column(JSONB, nullable=False)

    # Color for visual distinction
    color: Mapped[str] = mapped_column(String(7), nullable=False, default="#FFEB3B")

    # Tags
    tags: Mapped[list[str]] = mapped_column(
        ARRAY(String), nullable=False, server_default="{}"
    )

    # Relationships
    paper: Mapped["Paper"] = relationship("Paper")
    user: Mapped["User"] = relationship("User")

    def __repr__(self) -> str:
        return f"<PaperHighlight paper={self.paper_id}>"


class PaperLink(BaseModel):
    """Links between papers and other entities (projects, tasks, documents)."""

    __tablename__ = "paper_links"

    paper_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("papers.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # Link target (polymorphic)
    linked_entity_type: Mapped[str] = mapped_column(
        String(50), nullable=False
    )  # project, task, document
    linked_entity_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True), nullable=False, index=True
    )

    # Link metadata
    link_type: Mapped[str] = mapped_column(
        String(50), nullable=False, default="reference"
    )  # reference, citation, related
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Who created the link
    created_by_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )

    # Relationships
    paper: Mapped["Paper"] = relationship("Paper")
    created_by: Mapped["User | None"] = relationship("User")

    def __repr__(self) -> str:
        return f"<PaperLink paper={self.paper_id} -> {self.linked_entity_type}={self.linked_entity_id}>"
